rust/foxglove/src/websocket/protocol/client.rs

```diff
@@ -2,9 +2,128 @@
 //! Serializations are derived for testing.
 
 use crate::channel::ChannelId;
+use bytes::{Buf, Bytes};
 use serde::{Deserialize, Serialize};
+use tokio_tungstenite::tungstenite::Message;
 
-type Error = Box<dyn std::error::Error>;
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum ParseError {
+    #[error("Unexpected message type")]
+    UnexpectedMessageType,
+    #[error("Unknown binary opcode {0}")]
+    InvalidOpcode(u8),
+    #[error("Buffer too short")]
+    BufferTooShort,
+    #[error(transparent)]
+    Utf8(#[from] std::str::Utf8Error),
+    #[error(transparent)]
+    Json(#[from] serde_json::Error),
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum ClientMessage {
+    Subscribe(Subscribe),
+    Unsubscribe(Unsubscribe),
+    Advertise(ClientAdvertise),
+    Unadvertise(ClientUnadvertise),
+    MessageData(ClientMessageData),
+    GetParameters(GetParameters),
+    SetParameters(SetParameters),
+    SubscribeParameterUpdates(ParameterNames),
+    UnsubscribeParameterUpdates(ParameterNames),
+    ServiceCallRequest(ServiceCallRequest),
+    SubscribeConnectionGraph,
+    UnsubscribeConnectionGraph,
+    FetchAsset(FetchAsset),
+}
+impl ClientMessage {
+    pub fn op(&self) -> &'static str {
+        match self {
+            ClientMessage::Subscribe(_) => "subscribe",
+            ClientMessage::Unsubscribe(_) => "unsubscribe",
+            ClientMessage::Advertise(_) => "advertise",
+            ClientMessage::Unadvertise(_) => "unadvertise",
+            ClientMessage::MessageData(_) => "messageData",
+            ClientMessage::GetParameters(_) => "getParameters",
+            ClientMessage::SetParameters(_) => "setParameters",
+            ClientMessage::SubscribeParameterUpdates(_) => "subscribeParameterUpdates",
+            ClientMessage::UnsubscribeParameterUpdates(_) => "unsubscribeParameterUpdates",
+            ClientMessage::ServiceCallRequest(_) => "serviceCallRequest",
+            ClientMessage::SubscribeConnectionGraph => "subscribeConnectionGraph",
+            ClientMessage::UnsubscribeConnectionGraph => "unsubscribeConnectionGraph",
+            ClientMessage::FetchAsset(_) => "fetchAsset",
+        }
+    }
+
+    pub fn parse_message(message: Message) -> Result<Option<Self>, ParseError> {
```

Comment:
I think the original reason this was in websocket.rs was to keep the websocket Message type and logic segregated. I don't think that's particularly important, and I'm fine with doing it this way. I'm just offering that as context.